Quarkus-security - there is no way of selecting providers from extensions

### Description

I'm working on https://github.com/apache/camel-quarkus/issues/2005. I'm trying to use quarkus-security from camel-quarkus extensions.

I'm facing a problem, when extensions knows, that bouncycastle provider has to be used. This could be achieved by config property `quarkus.security.security-providers=BC`. Unfortunatelly I haven't found a way how to define it in an extension at build time. The only possible solution (which I see) is to force users, to define this property in their apps.

### Implementation ideas

I can imagine several ways of providing this new configuration option. 
For example:

- Existing build item `BouncyCastleProviderBuildItem` could be changed to allow multiple produce points, which would allow to configure it in extension.
- New build item could be created, which would allow definition of provider.

In both cases this new configuration could be validated against property.

Hi @sberyozkin I originally used the same approach as you are suggesting. I tried it again (now) with 

```
@BuildStep
    void produceBouncyCastleProvider(BuildProducer<BouncyCastleProviderBuildItem> bouncyCastleProvider) {
        bouncyCastleProvider.produce(new BouncyCastleProviderBuildItem());
    }
````

The result is an error:

```
Caused by: java.lang.RuntimeException: io.quarkus.builder.ChainBuildException: Multiple producers of item class io.quarkus.security.deployment.BouncyCastleProviderBuildItem (io.quarkus.security.deployment.SecurityProcessor#produceJcaSecurityProviders)

...
Caused by: java.lang.Throwable: This is the location of the conflicting producer (org.apache.camel.quarkus.component.crypto.deployment.CryptoProcessor#produceBouncyCastleProvider)
```
I think that problem is caused by the fact, that build item `extends SimpleBuildItem` and `SecurityProcessor` accepts `Optional<BouncyCastleProviderBuildItem> bouncyCastleProvider` and nt list.

--- a/extensions/security/deployment/pom.xml
+++ b/extensions/security/deployment/pom.xml
@@ -44,6 +44,11 @@
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-security-test-utils</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.bouncycastle</groupId>
+            <artifactId>bcprov-jdk15on</artifactId>
+            <scope>test</scope>
+        </dependency>
     </dependencies>
 
     <build>

--- a/extensions/security/deployment/src/main/java/io/quarkus/security/deployment/BouncyCastleJsseProviderBuildItem.java
+++ b/extensions/security/deployment/src/main/java/io/quarkus/security/deployment/BouncyCastleJsseProviderBuildItem.java
@@ -1,8 +1,10 @@
 package io.quarkus.security.deployment;
 
-import io.quarkus.builder.item.SimpleBuildItem;
+import java.util.Objects;
 
-public final class BouncyCastleJsseProviderBuildItem extends SimpleBuildItem {
+import io.quarkus.builder.item.MultiBuildItem;
+
+public final class BouncyCastleJsseProviderBuildItem extends MultiBuildItem {
     private final boolean inFipsMode;
 
     public BouncyCastleJsseProviderBuildItem() {
@@ -17,4 +19,18 @@ public final class BouncyCastleJsseProviderBuildItem extends SimpleBuildItem {
         return inFipsMode;
     }
 
+    @Override
+    public boolean equals(Object o) {
+        if (this == o)
+            return true;
+        if (o == null || getClass() != o.getClass())
+            return false;
+        BouncyCastleJsseProviderBuildItem that = (BouncyCastleJsseProviderBuildItem) o;
+        return inFipsMode == that.inFipsMode;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(inFipsMode);
+    }
 }

--- a/extensions/security/deployment/src/main/java/io/quarkus/security/deployment/BouncyCastleProviderBuildItem.java
+++ b/extensions/security/deployment/src/main/java/io/quarkus/security/deployment/BouncyCastleProviderBuildItem.java
@@ -1,8 +1,10 @@
 package io.quarkus.security.deployment;
 
-import io.quarkus.builder.item.SimpleBuildItem;
+import java.util.Objects;
 
-public final class BouncyCastleProviderBuildItem extends SimpleBuildItem {
+import io.quarkus.builder.item.MultiBuildItem;
+
+public final class BouncyCastleProviderBuildItem extends MultiBuildItem {
     private final boolean inFipsMode;
 
     public BouncyCastleProviderBuildItem() {
@@ -15,5 +17,20 @@ public final class BouncyCastleProviderBuildItem extends SimpleBuildItem {
 
     public boolean isInFipsMode() {
         return inFipsMode;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o)
+            return true;
+        if (o == null || getClass() != o.getClass())
+            return false;
+        BouncyCastleProviderBuildItem that = (BouncyCastleProviderBuildItem) o;
+        return inFipsMode == that.inFipsMode;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(inFipsMode);
     }
 }

--- a/extensions/security/deployment/src/main/java/io/quarkus/security/deployment/SecurityProcessor.java
+++ b/extensions/security/deployment/src/main/java/io/quarkus/security/deployment/SecurityProcessor.java
@@ -33,6 +33,7 @@ import io.quarkus.arc.deployment.AnnotationsTransformerBuildItem;
 import io.quarkus.arc.deployment.BeanArchiveIndexBuildItem;
 import io.quarkus.arc.deployment.InterceptorBindingRegistrarBuildItem;
 import io.quarkus.arc.deployment.SyntheticBeanBuildItem;
+import io.quarkus.builder.item.MultiBuildItem;
 import io.quarkus.deployment.Feature;
 import io.quarkus.deployment.annotations.BuildProducer;
 import io.quarkus.deployment.annotations.BuildStep;
@@ -119,8 +120,9 @@ public class SecurityProcessor {
     @BuildStep
     void prepareBouncyCastleProviders(BuildProducer<ReflectiveClassBuildItem> reflection,
             BuildProducer<RuntimeReinitializedClassBuildItem> runtimeReInitialized,
-            Optional<BouncyCastleProviderBuildItem> bouncyCastleProvider,
-            Optional<BouncyCastleJsseProviderBuildItem> bouncyCastleJsseProvider) throws Exception {
+            List<BouncyCastleProviderBuildItem> bouncyCastleProviders,
+            List<BouncyCastleJsseProviderBuildItem> bouncyCastleJsseProviders) throws Exception {
+        Optional<BouncyCastleJsseProviderBuildItem> bouncyCastleJsseProvider = getOne(bouncyCastleJsseProviders);
         if (bouncyCastleJsseProvider.isPresent()) {
             reflection.produce(
                     new ReflectiveClassBuildItem(true, true, SecurityProviderUtils.BOUNCYCASTLE_JSSE_PROVIDER_CLASS_NAME));
@@ -130,8 +132,11 @@ public class SecurityProcessor {
                     .produce(new RuntimeReinitializedClassBuildItem(
                             "org.bouncycastle.jsse.provider.DefaultSSLContextSpi$LazyManagers"));
             prepareBouncyCastleProvider(reflection, runtimeReInitialized, bouncyCastleJsseProvider.get().isInFipsMode());
-        } else if (bouncyCastleProvider.isPresent()) {
-            prepareBouncyCastleProvider(reflection, runtimeReInitialized, bouncyCastleProvider.get().isInFipsMode());
+        } else {
+            Optional<BouncyCastleProviderBuildItem> bouncyCastleProvider = getOne(bouncyCastleProviders);
+            if (bouncyCastleProvider.isPresent()) {
+                prepareBouncyCastleProvider(reflection, runtimeReInitialized, bouncyCastleProvider.get().isInFipsMode());
+            }
         }
     }
 
@@ -191,23 +196,28 @@ public class SecurityProcessor {
     @BuildStep
     @Record(ExecutionTime.STATIC_INIT)
     void recordBouncyCastleProviders(SecurityProviderRecorder recorder,
-            Optional<BouncyCastleProviderBuildItem> bouncyCastleProvider,
-            Optional<BouncyCastleJsseProviderBuildItem> bouncyCastleJsseProvider) {
+            List<BouncyCastleProviderBuildItem> bouncyCastleProviders,
+            List<BouncyCastleJsseProviderBuildItem> bouncyCastleJsseProviders) {
+        Optional<BouncyCastleJsseProviderBuildItem> bouncyCastleJsseProvider = getOne(bouncyCastleJsseProviders);
         if (bouncyCastleJsseProvider.isPresent()) {
             if (bouncyCastleJsseProvider.get().isInFipsMode()) {
                 recorder.addBouncyCastleFipsJsseProvider();
             } else {
                 recorder.addBouncyCastleJsseProvider();
             }
-        } else if (bouncyCastleProvider.isPresent()) {
-            recorder.addBouncyCastleProvider(bouncyCastleProvider.get().isInFipsMode());
+        } else {
+            Optional<BouncyCastleProviderBuildItem> bouncyCastleProvider = getOne(bouncyCastleProviders);
+            if (bouncyCastleProvider.isPresent()) {
+                recorder.addBouncyCastleProvider(bouncyCastleProvider.get().isInFipsMode());
+            }
         }
     }
 
     @BuildStep
     void addBouncyCastleProvidersToNativeImage(BuildProducer<NativeImageSecurityProviderBuildItem> additionalProviders,
-            Optional<BouncyCastleProviderBuildItem> bouncyCastleProvider,
-            Optional<BouncyCastleJsseProviderBuildItem> bouncyCastleJsseProvider) {
+            List<BouncyCastleProviderBuildItem> bouncyCastleProviders,
+            List<BouncyCastleJsseProviderBuildItem> bouncyCastleJsseProviders) {
+        Optional<BouncyCastleJsseProviderBuildItem> bouncyCastleJsseProvider = getOne(bouncyCastleJsseProviders);
         if (bouncyCastleJsseProvider.isPresent()) {
             additionalProviders.produce(
                     new NativeImageSecurityProviderBuildItem(SecurityProviderUtils.BOUNCYCASTLE_JSSE_PROVIDER_CLASS_NAME));
@@ -215,12 +225,22 @@ public class SecurityProcessor {
                     ? SecurityProviderUtils.BOUNCYCASTLE_FIPS_PROVIDER_CLASS_NAME
                     : SecurityProviderUtils.BOUNCYCASTLE_PROVIDER_CLASS_NAME;
             additionalProviders.produce(new NativeImageSecurityProviderBuildItem(providerName));
-        } else if (bouncyCastleProvider.isPresent()) {
-            final String providerName = bouncyCastleProvider.get().isInFipsMode()
-                    ? SecurityProviderUtils.BOUNCYCASTLE_FIPS_PROVIDER_CLASS_NAME
-                    : SecurityProviderUtils.BOUNCYCASTLE_PROVIDER_CLASS_NAME;
-            additionalProviders.produce(new NativeImageSecurityProviderBuildItem(providerName));
+        } else {
+            Optional<BouncyCastleProviderBuildItem> bouncyCastleProvider = getOne(bouncyCastleProviders);
+            if (bouncyCastleProvider.isPresent()) {
+                final String providerName = bouncyCastleProvider.get().isInFipsMode()
+                        ? SecurityProviderUtils.BOUNCYCASTLE_FIPS_PROVIDER_CLASS_NAME
+                        : SecurityProviderUtils.BOUNCYCASTLE_PROVIDER_CLASS_NAME;
+                additionalProviders.produce(new NativeImageSecurityProviderBuildItem(providerName));
+            }
         }
+    }
+
+    private <BI extends MultiBuildItem> Optional<BI> getOne(List<BI> items) {
+        if (items.size() > 1) {
+            throw new IllegalStateException("Only a single Bouncy Castle registration can be provided.");
+        }
+        return items.stream().findFirst();
     }
 
     /**

--- a/extensions/security/deployment/src/test/java/io/quarkus/security/test/DuplicatedDifferentBouncycastleProviderTest.java
+++ b/extensions/security/deployment/src/test/java/io/quarkus/security/test/DuplicatedDifferentBouncycastleProviderTest.java
@@ -1,0 +1,22 @@
+package io.quarkus.security.test;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+
+public class DuplicatedDifferentBouncycastleProviderTest {
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .setExpectedException(IllegalStateException.class)
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
+                    .addAsResource("application-duplicated-different-providers.properties", "application.properties"));
+
+    @Test
+    void shouldThrow() {
+        Assertions.fail("An IllegalStateException should have been thrown due to duplicated bouncycastle providers");
+    }
+}

--- a/extensions/security/deployment/src/test/java/io/quarkus/security/test/DuplicatedTheSameBouncycastleProviderTest.java
+++ b/extensions/security/deployment/src/test/java/io/quarkus/security/test/DuplicatedTheSameBouncycastleProviderTest.java
@@ -1,0 +1,20 @@
+package io.quarkus.security.test;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+
+public class DuplicatedTheSameBouncycastleProviderTest {
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
+                    .addAsResource("application-duplicated-the-same-providers.properties", "application.properties"));
+
+    @Test
+    public void testSuccess() {
+        //execution of the test is enough to confirm that configuration is allowed
+    }
+}

--- a/extensions/security/deployment/src/test/resources/application-duplicated-different-providers.properties
+++ b/extensions/security/deployment/src/test/resources/application-duplicated-different-providers.properties
@@ -1,0 +1,1 @@
+quarkus.security.security-providers=BC,BCFIPS

--- a/extensions/security/deployment/src/test/resources/application-duplicated-the-same-providers.properties
+++ b/extensions/security/deployment/src/test/resources/application-duplicated-the-same-providers.properties
@@ -1,0 +1,1 @@
+quarkus.security.security-providers=BC,BC

